### ISSUE TITLE
Blockstate Argument

### ISSUE THREAD

**Issue by SpaceCheetah** [JorelAli/1.13-Command-API#90]
In some vannilla commands, such as setblock, there is a way to enter a block with data (and nbt, though as far as I know there isn't really a single class in Spiggot that has both the blockdata and the tile entity). In the command registration, it is referred to as "minecraft:block_state". You can already get a blockstate from a string, with Bukkit.createBlockState(string), that functions the same way as setblock except the nbt part. However, I could neither figure out a way to make it turn red if invalid, nor give suggestions, for the data part (I could for the material part using a literal argument, but that seems inefficient).
Example:
Input: minecraft:oak_log
Output: The default BlockData for an oak log.
Example:
Input: minecraft:oak_log[axis=z]
Output: A BlockData for an oak fence with the axis data set to z.
As typing it should show suggestions the same way it does when typing the block argument in commands such as setblock and fill, and turn the argument red when the input is invalid.

This is a repost, as the account I originally posted this on was flagged and marked as hidden for the public (probably because I used a temp email; I don't like sharing my real email with websites). Anyway that means you probably don't see the original post, so I'm just reposting it.

**Comment by Minenash** [comment]
For the record, this was denied in this issue:
https://github.com/JorelAli/1.13-Command-API/issues/69#issuecomment-525006172

***
`minecraft:block_state` and `minecraft:block_predicate` - These arguments have rather complicated internals compared to other arguments, which I don't believe would fit in well with Bukkit at all. Therefore no.
***

However, there were others that were denied in that comment, and then added anyways. From looking at the argument, it would need to be able to accept nbt, so it would need the nbt library thing. This is already done for nbt, so I don't think that's a huge issue. As for implementation, it could possibly return both the object created from the method you mentioned, and the nbt object created by the nbt library. I don't know anything about the nbt library, but if it has a way to take in a block and apply the nbt, then it could return that.

**Comment by JorelAli** [comment]
> Example:
Input: minecraft:oak_log
Output: The default BlockData for an oak log.
Example:
Input: minecraft:oak_log[axis=z]
Output: A BlockData for an oak fence with the axis data set to z.

This is a surprisingly good example. This example alone may be enough to be worth investigating the `minecraft:block_state` implementation and how it may fit into the CommandAPI. If I recall correctly, the internal implementation was very complicated and trying to get it to cooperate with Bukkit was difficult. Despite that, it doesn't mean it's impossible. Some new developments have been made (despite the fact that the current project status is basically "on hold"), and trying to "build the bridge" between vanilla and Bukkit may be doable.

In short, it's a good example - at some point later on I'll revisit my decision to add it as an argument. Since the project is "on hold" according to [this](https://github.com/JorelAli/1.13-Command-API#psa-v30-release-date), no further action will be taken until whenever.

**Comment by SpaceCheetah** [comment]
I'll look into seeing if I can get it to work myself; I haven't done much with reflection yet though, so I have quite a bit to learn.

**Comment by JorelAli** [comment]
So basically, in order to use the `minecraft:block_state` tag, you _have_ to have an actual physical block in the world. In order to refer to this block, the exact coordinates of this block also have to be passed to the block argument.

Minecraft's commands have some good examples of what I mean:

- `/setblock <pos> <block_state>` 
- `/fill<from> <to> <block_state>` 

Both of these commands require some input of position in order to be used. However, the `ItemStackArgument` lets you apply NBT (I believe?) to the item, and the item is given to the player. This seems like the best way to handle this sort of thing. The CommandAPI cannot guarantee that a position would be "provided" to the `block_state` argument, and it's already difficult to do so given its current implementation.

-----

For example, say we have the following command declaration:

```java
LinkedHashMap<String, Argument> args = new LinkedHashMap<>();
args.put("pos", new LocationArgument());
args.put("block", new BlockStateArgument());
```

Ideally, in order to pass the position to the block state argument, you'd want something along the lines of:

```java
LinkedHashMap<String, Argument> args = new LinkedHashMap<>();
args.put("pos", new LocationArgument());
args.put("block", new BlockStateArgument(/* Some position here */));
```

However, perhaps you want to handle it at execution time (For example, that fill command):

```java
new CommandAPICommand("cmd").withArguments(args).executes((s, a) -> {
    ?? blockState = (??) a[0];
    blockState.withPos(/* Some position here */);
    blockState.apply(); //Or something like this
}).register();
```

Now, this _is_ technically a solution, but I don't like it in the slightest. The CommandAPI tries to be as close to regular Bukkit as it can be, and this seems very non-Bukkit like. It's impossible to use `apply()` without previously setting the position, and I don't want to permit the potential for errors to occur.

-----

To summarize, I'm still siding with the "don't implement it" choice at the moment, for the following reasons:

- It introduces the risk of errors (if a developer doesn't provide a position to the blockstate, it's undefined)
- From an implementation point of view, it's still very very complicated. The block's NBT data and regular data (such as its type, orientation, position, which is undefined until declared using some hackery) are separated. The process of converting this complicated NMS block information into a Bukkit `Block` object is still unknown.
- It requires another class (denoted as `??` in the example above), which moves away from the Bukkit-like API that developers are familiar with. This is also the same reason I was reluctant to implement a better form of the `MathOperation` argument.

**Comment by 469512345** [comment]
What's wrong with an implementation like this, using Bukkit's own BlockData object? I have played around with this on my own and it works flawlessly. It's nice because setting the blockdata automatically updates the material of the block too so it really is as simple as that
//Code in the Versioned NMS class
```java
public BlockData getBlockData(CommandContext cmdCtx, String str) {
	return CraftBlockData.fromData(ArgumentTile.a(cmdCtx, str).a());
}
```
//Code in the command executor
```java
Block block;
BlockData data = (BlockData) args[0];
block.setBlockData(data);
```

Surely something like this is close to regular bukkit, because well it is regular bukkit?

**Comment by JorelAli** [comment]
@469512345 
Nice research! I'll make sure to incorporate this into the CommandAPI, thanks very much!

**Comment by 469512345** [comment]
Happy to help

**Comment by JorelAli** [comment]
Implemented in this commit 0803e410af3f79bd2c91340c6ead4e9c51a4296c for version 3.1
